zipline/sources/test_source.py

```diff
@@ -153,10 +150,21 @@ def __init__(self, *args, **kwargs):
                 self.event_list[1].dt - self.event_list[0].dt)
             self.concurrent = kwargs.get('concurrent', False)
 
+            self.identifiers = kwargs.get(
+                'sids',
+                np.unique([event.sid for event in self.event_list]).tolist())
+            env.update_asset_finder(erase_existing=True,
+                                    identifiers=self.identifiers)
+            assets, _ = env.asset_finder.lookup_generic(self.identifiers,
+                                                        as_of_date=self.start)
+            self.sids = [asset.sid for asset in assets]
+            for event in self.event_list:
+                event.sid = env.asset_finder.lookup_generic(event.sid,
```

Comment:
I factored-out lookup_generic everywhere except TradingAlgorithm.symbol(). I made a new method retrieve_asset_by_identifier() that is used in its place here.